async-nats/src/jetstream/consumer/pull.rs

```diff
@@ -246,7 +252,110 @@ impl<'a> futures::Stream for Sequence<'a> {
     }
 }
 
-pub type Stream<'a> = stream::TryFlatten<Sequence<'a>>;
+pub struct Stream<'a> {
+    pending_messages: usize,
+    subscriber: Subscriber,
+    context: Context,
+    inbox: String,
+    subject: String,
+    batch_config: BatchConfig,
+    request: Option<BoxFuture<'a, Result<(), Error>>>,
+}
+
+impl<'a> Stream<'a> {
+    async fn stream(
+        batch_config: BatchConfig,
+        consumer: &Consumer<Config>,
+    ) -> Result<Stream<'a>, Error> {
+        let inbox = consumer.context.client.new_inbox();
+        let subscription = consumer.context.client.subscribe(inbox.clone()).await?;
+        let subject = format!(
+            "{}.CONSUMER.MSG.NEXT.{}.{}",
+            consumer.context.prefix, consumer.info.stream_name, consumer.info.name
+        );
+
+        Ok(Stream {
+            pending_messages: 0,
+            subscriber: subscription,
+            context: consumer.context.clone(),
+            request: None,
+            inbox,
+            subject,
+            batch_config,
+        })
+    }
+}
+
+impl<'a> futures::Stream for Stream<'a> {
+    type Item = Result<jetstream::Message, Error>;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        match self.request.as_mut() {
+            None => {
+                let context = self.context.clone();
+                let inbox = self.inbox.clone();
+                let subject = self.subject.clone();
+
+                if self.pending_messages < std::cmp::min(self.batch_config.batch / 2, 100) {
+                    let batch = self.batch_config.batch;
+                    self.pending_messages += batch;
+                    self.request = Some(Box::pin(async move {
+                        let request = serde_json::to_vec(&BatchConfig {
+                            batch,
+                            ..Default::default()
+                        })
+                        .map(Bytes::from)?;
+
+                        context
+                            .client
+                            .publish_with_reply(subject, inbox, request)
+                            .await?;
+                        Ok(())
+                    }));
+                }
+
+                if let Some(request) = self.request.as_mut() {
+                    match request.as_mut().poll(cx) {
+                        Poll::Ready(result) => {
+                            self.request = None;
+                            result?;
+                        }
+                        Poll::Pending => {}
+                    }
+                }
+            }
+
+            Some(request) => match request.as_mut().poll(cx) {
```

Comment:
Not using the result from the poll, so failed requests are just silently thrown away.

```suggestion
            Some(request) => match request.as_mut().poll(cx) {
```

Reply:
as discussed - we are using the result.